### ISSUE TITLE
bz_detectBadChannels.m,  we need it.

### ISSUE THREAD

**Issue by DavidTingley** [buzsakilab/buzcode#28]

**Comment by brendonw1** [comment]
Ideas??
Maybe you wanna look for channels with very different power spectrum from the others?

One cheap idea is to have the user hide channels or remove channels or something in neuroscope and use that as a quick way to do it... by reading the nrs/xml files
It'd be a stopgap, but very helpful for timebeing

**Comment by DavidTingley** [comment]
raising this to our 'first release' goal so that we hopefully revisit in the next few months

**Comment by DavidTingley** [comment]
@brendonw1,  I think the 'very different' power spectrum would be a good first heuristic to remove completely messed up channels that are bad for the whole recording.  We'd maybe want to do this in a moving window fashion to find channels that "sometimes" drop out, more often than others.

**Comment by juneshuoyang** [comment]
Can you give me a dataset/recording session where a bad channel exists but not removed yet? Thanks.

**Comment by brendonw1** [comment]
June, will you try this one?

**Comment by brendonw1** [comment]
Or did you want me to do it?

**Comment by juneshuoyang** [comment]
Yes, I'd like to try it. Sam randomly opened a dataset which has a bad
channel. I think I sort of understand the situation.

What will be your method? Maybe you want to temporarily keep it secret, so
I can have a chance?

Since I am new to this field, I would ask 3 weeks to work on it.

Also, irregular pattern detection can always be improved to a certain
degree. Without a large amount of existing bad patterns, how we test this
function?

**Comment by dlevenstein** [comment]
Almost every recording has at least 1 bad channel, and a number of bad time epochs on good channels. These are two separate problems, but maybe they could be addressed with the same function?  This would be a huge deal for the lab :D

I'd be happy to help put together a test dataset with manually marked bad channels, and bad times to test such a function.

**Comment by DavidTingley** [comment]
@juneshuoyang, I'll also contribute recordings with good/bad channels marked (I can work with you on this tomorrow or the next time you're in).  I think this will be very easy for the first problem, identifying channels that are bad for an entire recording.  

The second problem is identifying channels that are transiently bad (disconnected, more susceptible to noise/muscle artifact). This makes me think we'll want some type of 'quality score' (i.e. 10=great,0=useless). 

The final problem is identifying intervals of time where ALL channels are bad.  There are a number of great recordings with 10-100 seconds of messed up data that screw up all of our detectors.  Maybe over time we can build an algorithm that does all three of these things, but it makes sense to focus on one at a time for now.

**Comment by juneshuoyang** [comment]
@DavidTingley sure, thanks! I am curious about your method if you think there is a very easy way to deal with the first problem.

**Comment by brendonw1** [comment]
David really well formulated. That's great.

For 1 and 2 I'd basically use amplitudes of vector projections of
individual spectra against mean spectra.

The problem 1 you'd take the average spectrum for each channel and project
that agains the average of all the spectra for the channels in the local
group (probe). Take the amplitudes of those projections as your basic
similarity metric and then try to find outliers. To do that I might just
make a histogram and look by eye to see what does/doesn't make sense.

As for problem 2 I'm trying to think if reasonably fast ways of doing this
but I'd think of something like taking a spectrogram (?5 second
resolution?...obviously this can be adjusted), take the mean spectrum and
then project each second against the mean for that channel.

Three points:
1) I'd sample at Log-spaced frequencies. Actually I might do something like
unique(round(logspace(0,log10(625),30)))

2) I'd multiply the power at each frequency by the frequency itself in
order to get rid of 1/f falloff - so that you don't under emphasize high
frequencies

3) for davids question 2 you might need to find a way to compensate for
slow wave sleep (especially if in cortex, amygdala etc). SWS may will
appear to be an aberration itself and could really mess things up. Maybe do
PCA and do like Dan L did in our paper to find times stronger/weaker in PC1
projection and then consider spectral changes separately for those seconds.
Another way is to somehow make sure that any changes you see in your
channel aren't matched in most neighbor channels. Like if neighbors all
suddenly deviate from their own mean spectra the same way at the same time
then you're not concerned (at least from a question 2 perspective).

As for question 3, maybe you do like that last part I just mentioned
looking at all channels at once and decide that it's ok if they all deviate
together but reassure yourself if that deviation is to some other commonly
found state within that recording. But you get concerned if the deviation
of many channels is to a really unusual state.

Sorry if that's too much.

**Comment by dlevenstein** [comment]
I wonder if we could incorporate a general denoising/artefact removal process.
would this be more appropriate in bz_LFPfromDat?

Saw this recently and thought it could be useful to incorporate in our pipeline:
http://www.brainwavelet.org/about/



Another thing that would be good is a "bz_findRepresentativeChannels" function, that picks the "best" channel from each region and saves them in a baseName.RepresentativeLFP.lfp.mat file. (For a more portable dataset? I've enjoyed having this on my laptop in the past).

**Comment by DavidTingley** [comment]
@dlevenstein, I think any denoising/manipulation should come later on in the pipeline, and generate a separate file/filetype.  bz_LFPFromDat (and all *.lfp) files should be as close to the raw data as possible.  

Best channel defined how? lowest noise? highest amplitude theta? highest amplitude ripple? anatomical region?  My solution for this has been to have a set of fields in the sessionInfo struct that identify single channels to be used for specific analyses
